pageserver/src/import_datadir.rs

```diff
@@ -620,6 +623,68 @@ async fn read_all_bytes(reader: &mut (impl AsyncRead + Unpin)) -> Result<Bytes>
     Ok(Bytes::from(buf))
 }
 
+const YIELD_DIST: u64 = 1024;
+
+struct YieldingVec {
+    yld_len: u64,
+    buf: Vec<u8>,
+}
+
+impl YieldingVec {
+    fn new() -> Self {
+        Self {
+            yld_len: 0,
+            buf: Vec::new(),
+        }
+    }
+    fn should_yield(&mut self, buf_len: usize) -> bool {
+        let buf_len: u64 = buf_len.try_into().unwrap();
+        let ret = self.yld_len / YIELD_DIST != (self.yld_len + buf_len) / YIELD_DIST;
+        self.yld_len += buf_len;
+        ret
+    }
+}
+
+impl AsyncWrite for YieldingVec {
+    fn poll_write(
+        mut self: Pin<&mut Self>,
+        _cx: &mut task::Context<'_>,
+        buf: &[u8],
+    ) -> Poll<std::io::Result<usize>> {
+        if self.should_yield(buf.len()) {
+            return Poll::Pending;
+        }
+        self.get_mut().buf.extend_from_slice(buf);
+        Poll::Ready(Ok(buf.len()))
+    }
+
+    fn poll_write_vectored(
+        mut self: Pin<&mut Self>,
+        _: &mut task::Context<'_>,
+        bufs: &[IoSlice<'_>],
+    ) -> Poll<std::io::Result<usize>> {
+        if self.should_yield(bufs.iter().map(|b| b.len()).sum()) {
+            return Poll::Pending;
+        }
+        Poll::Ready(std::io::Write::write_vectored(&mut self.buf, bufs))
+    }
+
+    fn is_write_vectored(&self) -> bool {
+        true
+    }
```

Comment:
If you'd leave these two out, would it not just forward to the one poll_write method? https://docs.rs/tokio/latest/src/tokio/io/async_write.rs.html#158-168

Reply:
removed